rfd/0229a-scoped-join-tokens.md

```diff
@@ -0,0 +1,259 @@
+
+   // Spec is the token specification.
+   ScopedTokenSpec spec = 6;
++  // The status of the token
++  ScopedTokenStatus status = 7;
+ }
+
+ // ScopedTokenSpec is the specification of a scoped token.
+ message ScopedTokenSpec {
+-  // AssignedScope is the scope to which this token is assigned.
++  // The scope to which this token is assigned.
+   string assigned_scope = 1;
+
+-  // TODO(fspmarshall): port relevant token features to scoped tokens.
++  // The list of roles associated with the token. They will be converted
++  // to metadata in the SSH and X509 certificates issued to the user of the
++  // token.
++  repeated string roles = 2;
++
++  // The joining method required in order to use this token.
++  // Supported joining methods for scoped tokens only include 'token'.
++  string join_method = 3;
++
++  // The set of labels to be automatically assigned to any provisioned SSH node.
++  map<string, string> ssh_labels = 4;
++
++  // The number of resources that can be provisioned using this token.
++  int32 max_uses = 5;
++}
++
++// The status of a scoped token.
++message ScopedTokenStatus {
++  // The number of successful provisioning attempts made using this token.
++  int32 attempted_uses = 6;
+ }
+```
+
+### Provisioning
+
+Scoped resources will be provisioned by extending the existing `Join` RPC to
+support scoped tokens. When the join method is `token` the auth server will
+query both the existing `ProvisionTokenV2` and the new `ScopedToken` resources
+using the provided token name. Conflicting token names between scoped and
+unscoped tokens will result in an error and it will be up to the administrator
+to resolve the ambiguity. Either by deleting one of the conflicting tokens or
+creating a new token with a different name. This approach allows for backwards
+compatible provisioning where agents can be scoped without any knowledge of
+scopes themselves. The `assigned_scope` of the `ScopedToken` will be assigned
+as the scope of the resulting resource. It will also be attached as metadata to
+the resulting host certificate under the `AgentScope` field so that scope
+related access controls can be performed against the agent's identity rather
+than just restricting clients' access to the agent itself.
+
+### Automatic labels for SSH nodes
+
+Scoped tokens should also support automatic assignment of labels to any SSH
+node provisioned. This document only proposes support for SSH nodes as solving
+for all scoped node types introduces more significant complexity. The
+`ssh_labels` field of a `ScopedToken` will be added encoded into the resulting
+host certificate in much the same way as the `assigned_scope`. These labels
+will be extracted while registering the inventory control stream and applied as
+a new `token_labels` field stored in `ServerSpecV2`. Future heartbeats will
+verify `token_labels` just as they do for static labels and command labels.
+
+Merging token-assigned labels with static labels was also considered as a way
```

Comment:
What is the priority of label sources? I assume token-assigned labels will not be allowed to be overwritten by static or commands labels but it would be good to explicitly call that out.